add_docstrings.py

```diff
@@ -155,31 +310,62 @@ def leave_ClassDef(
             )
 
     def object_fullname(self, final_part: str) -> str:
+        """Given an unqualified name, convert it to a fully qualified name.
+
+        For example, given the name `method`, if we are visiting a class `Foo`
+        inside a class `Bar` inside a module `spam`, this will return `spam.Bar.Foo.method`.
+
+        If the name is mangled, it will be mangled according to the current class
+        namespace: if we were given the name `__method`, we would instead return
+        `spam.Bar.Foo._Foo__method`.
+        """
         parent_fullname = ".".join(parent.name for parent in self.runtime_parents)
         mangled_name = self.maybe_mangled_name(final_part)
         return f"{parent_fullname}.{mangled_name}"
 
-    def log_runtime_object_not_found(self, name: str) -> None:
+    def _log_runtime_object_not_found(self, name: str) -> None:
         log(f"Could not find {self.object_fullname(name)} at runtime")
 
+    @override
     def leave_FunctionDef(
         self, original_node: libcst.FunctionDef, updated_node: libcst.FunctionDef
     ) -> libcst.FunctionDef:
+        """Attempt to add a docstring to the function definition."""
+
         runtime_parent = self.runtime_parents[-1]
         if runtime_parent.value.is_not_found():
             return original_node
         runtime_object = get_runtime_object_for_stub(
-            runtime_parent=runtime_parent,
             name=self.maybe_mangled_name(updated_node.name.value),
+            runtime_parent=runtime_parent,
         )
         if runtime_object is None:
-            self.log_runtime_object_not_found(updated_node.name.value)
+            self._log_runtime_object_not_found(updated_node.name.value)
             return original_node
         return self.document_class_or_function(
             updated_node=updated_node, runtime_object=runtime_object
         )
 
     def _leave_Suite(self, original_node: SuiteT, updated_node: SuiteT) -> SuiteT:
+        """Remove unwanted changes that may have been made to function definitions inside a "suite".
+
+        A "suite" is any namespace that can contain multiple statements: in practice, this means
+        a module namespace or a class namespace.
+
+        `leave_FunctionDef` will add docstrings to all function definitions where it can find
+        a corresponding runtime object. But it isn't always desirable to add a docstring to
+        *every* function definition in a suite. In particular:
+
+          - If there are multiple overloads for a function in a suite, it's redundant to add
+            the docstring to every overload. Just adding a docstring to the first overload in
+            the suite will be sufficient.
+          - If the suite contains a property that has a setter, we don't need to add the
+            property's docstring to the setter as well as the getter.
+
+        This method therefore iterates through the body of the suite and removes any changes
```

Comment:
Could be worth mentioning why you do it this way, rather than not adding the bad docstrings in the first place.

Reply:
it basically comes down to: it was easy to figure out how to do it this way, and I couldn't immediately figure out how to do it the way you suggest 😆

Reply:
added a note in https://github.com/astral-sh/docstring-adder/pull/16/commits/0b0073c23bc96b158507de9beb7917e482c9383a